pacman/package.go

```diff
@@ -0,0 +1,329 @@
+package pacman
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/rs/zerolog/log"
+	"github.com/schollz/progressbar/v3"
+
+	"github.com/jfhbrook/dosapp/config"
+)
+
+type Package struct {
+	Name                   string
+	remotePackageExists    bool
+	LocalVersion           *semver.Version
+	LocalReleaseVersion    *semver.Version
+	UpstreamVersion        *semver.Version
+	UpstreamReleaseVersion *semver.Version
+	URL                    string
+	Stage                  *Stage
+	Config                 *config.Config
+}
+
+func localPackagePath(name string, conf *config.Config) string {
+	return filepath.Join(conf.PackageHome, name)
+}
+
+// TODO: This call signature is great for a remote package, but many of my
+// use cases are for packages only installed locally - and those are having
+// to inject a bunch of null values.
+func NewPackage(
+	name string,
+	version *semver.Version,
+	releaseVersion *semver.Version,
+	url string,
+	ch *Stage,
+	conf *config.Config,
+) *Package {
+	packagePath := filepath.Join(localPackagePath(name, conf), "Package.yml")
+
+	pkg := &Package{}
+	pkg.Name = name
+	pkg.Config = conf
+
+	pkg.Stage = ch
+	pkg.remotePackageExists = version != nil && releaseVersion != nil && url != ""
+	pkg.UpstreamVersion = version
+	pkg.UpstreamReleaseVersion = releaseVersion
+	pkg.URL = url
+
+	m, err := ManifestFromFile(packagePath)
+
+	if err != nil {
+		log.Debug().Str("path", packagePath).Err(err).Msg("Installed package not found")
+	} else {
+		pkg.LocalVersion = m.Version
+		pkg.LocalReleaseVersion = m.ReleaseVersion
+	}
+	return pkg
+}
+
+func (pkg *Package) RemotePackageExists() bool {
+	return pkg.remotePackageExists
+}
+
+func (pkg *Package) LocalArtifactPath() string {
+	return filepath.Join(
+		pkg.Config.ArtifactHome,
+		fmt.Sprintf(
+			"%s-%s-%s.tar.gz",
+			pkg.Name,
+			pkg.UpstreamVersion.String(),
+			pkg.UpstreamReleaseVersion.String(),
+		),
+	)
+}
+
+func (pkg *Package) MkArtifactDir() error {
```

Comment:
I suspect some of these methods are dead code. But I'm hesitant to whack-a-mole them at this time. This PR was already a big lift.